### PR TITLE
Improvement: Reader: Add sorting option to Discover or Topics

### PR DESCRIPTION
**Pull request by KaungMyat-Min**
# Feature
add application layer for reader sorting feature

## To do:
- implement ui components
- add unit tests for changing sorting types

## To test
- verify the pre-defined unit tests pass

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt [hunk 1]
```diff
@@ -2,28 +2,41 @@ package org.wordpress.android.datasets
...
+    }
+
     fun dropTables(db: SQLiteDatabase) {
         db.execSQL("DROP TABLE IF EXISTS tbl_discover_cards")
     }
 
-    fun clear() {
-        AppLog.i(AppLog.T.READER, "clearing ReaderDiscoverCardsTable")
-        getWritableDb().delete(DISCOVER_CARDS_TABLE, null, null)
+    fun clear(DiscoverSortingType: DiscoverSortingType) {
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt]: Minor 🔍: Typo, `DiscoverSortingType` variable name should be camelCased.
Same applies on line 50
`fun addCardsPage(cardsJson: String, DiscoverSortingType: DiscoverSortingType) {`

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt [hunk 2]
```diff
@@ -2,28 +2,41 @@ package org.wordpress.android.datasets
...
     private const val DISCOVER_CARDS_TABLE = "tbl_discover_cards"
     private const val CARDS_JSON_COLUMN = "cards_json"
+    private const val CARDS_SORTING_TYPE_COLUMN = "cards_sorting_type"
+
     fun createTable(db: SQLiteDatabase) {
         db.execSQL(
                 "CREATE TABLE IF NOT EXISTS $DISCOVER_CARDS_TABLE (" +
                         "  _id INTEGER PRIMARY KEY AUTOINCREMENT," +
                         " $CARDS_JSON_COLUMN TEXT" +
+                        " $CARDS_SORTING_TYPE_COLUMN TEXT" +
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt]: Warning ⚠️: This column will not be added to the table due to missing comma on the previous line.

App crashes with the logs: 
```
android.database.sqlite.SQLiteException: no such column: cards_sorting_type (code 1 SQLITE_ERROR): , while compiling: SELECT * FROM tbl_discover_cards WHERE cards_sorting_type = ? ORDER BY _id ASC
```

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt [hunk 3]
```diff
@@ -2,28 +2,41 @@ package org.wordpress.android.datasets
...
     }
 
-    fun clear() {
-        AppLog.i(AppLog.T.READER, "clearing ReaderDiscoverCardsTable")
-        getWritableDb().delete(DISCOVER_CARDS_TABLE, null, null)
+    fun clear(DiscoverSortingType: DiscoverSortingType) {
+        val args = arrayOf(DiscoverSortingType.sortedBy)
+        val whereClause = "WHERE $CARDS_SORTING_TYPE_COLUMN = ?" +
+                AppLog.i(AppLog.T.READER, "clearing ReaderDiscoverCardsTable")
+        getWritableDb().delete(DISCOVER_CARDS_TABLE, whereClause, args)
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt]: Warning ⚠️:  Delete query translates to 

```
DELETE FROM tbl_discover_cards WHERE WHERE cards_sorting_type = ?kotlin.Unit      
```

with duplicate `WHERE`.  Also log statement gets appended to where clause resulting in a `kotlin.Unit` addition at the end.

Crash logs:

```
FATAL EXCEPTION: DefaultDispatcher-worker-1
    Process: org.wordpress.android.prealpha, PID: 28900
    android.database.sqlite.SQLiteException: near "kotlin": syntax error (code 1 SQLITE_ERROR): , while compiling: DELETE FROM tbl_discover_cards WHERE cards_sorting_type = ?kotlin.Unit
```

File: WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverViewModel.kt [hunk 4]
```diff
@@ -88,6 +91,9 @@ class ReaderDiscoverViewModel @Inject constructor(
     private val _preloadPostEvents = MediatorLiveData<Event<PreLoadPostContent>>()
     val preloadPostEvents: LiveData<Event<PreLoadPostContent>> = _preloadPostEvents
 
+    private val _sortingType = MutableLiveData<DiscoverSortingType>(POPULARITY)
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverViewModel.kt]: Minor 🔍: (IDE warning) Explicit type argument `DiscoverSortingType` can be removed .

```suggestion
    private val _sortingType = MutableLiveData(DiscoverSortingType.POPULARITY)
```

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt [hunk 5]
```diff
@@ -263,20 +294,20 @@ class ReaderDiscoverLogic(
...
     }
 
-    private suspend fun clearCache() {
-        val blogIds = getRecommendedBlogsToBeDeleted().map { it.blogId }
+    private suspend fun clearCache(sortingType: DiscoverSortingType) {
+        val blogIds = getRecommendedBlogsToBeDeleted(sortingType).map { it.blogId }
         ReaderBlogTable.deleteBlogsWithIds(blogIds)
 
-        ReaderDiscoverCardsTable.clear()
+        ReaderDiscoverCardsTable.clear(sortingType)
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt:309]: Question ❓: Here post ids for only selected `sortingType` will be cleared from `ReaderDiscoverCardsTable` but _all_ discover posts will be deleted from the `ReaderPostTable` in the following line:
`ReaderPostTable.deletePostsWithTag(ReaderTag.createDiscoverPostCardsTag())`

Are we aiming to store post ids for each sort type in the `ReaderDiscoverCardsTable` or just the last one?

File: WordPress/src/main/java/org/wordpress/android/util/config/FilterDiscoverFeatureConfig.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.wordpress.android.BuildConfig
+import org.wordpress.android.annotation.Feature
+import org.wordpress.android.util.config.FilterDiscoverFeatureConfig.Companion.FILTER_DISCOVER
+import javax.inject.Inject
+
+/**
+ * Configuration of the filter for discover or topic feature.
+ */
+@Feature(remoteField = FILTER_DISCOVER, defaultValue = true)
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/util/config/FilterDiscoverFeatureConfig.kt]: There are two sections "Remote features" and "Features in development" in the "Test feature configuration" screen. We need to use `@FeatureInDevelopment` annotation here similar to the sample code I shared. 

Apologies, I didn't make this clear, I understand it is difficult to notice without context.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt [hunk 7]
```diff
@@ -20,18 +21,22 @@ import org.wordpress.android.util.AppLog.T.READER
...
             intent.putExtra(ARG_DISCOVER_TASK, task)
+            intent.putExtra(ARG_SORTING_TYPE, sortingType)
             context.startService(intent)
         } else {
             // schedule the JobService here for API >= 26. The JobScheduler is available since API 21, but
             // it's preferable to use it only since enforcement in API 26 to not break any old behavior
             val componentName = ComponentName(context, ReaderDiscoverJobService::class.java)
             val extras = PersistableBundle()
             extras.putInt(ARG_DISCOVER_TASK, task.ordinal)
+            extras.putInt(ARG_SORTING_TYPE, task.ordinal)
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt]: Warning ⚠️: Passing ordinal of a wrong enum:

```suggestion
            extras.putInt(ARG_SORTING_TYPE, sortingType.ordinal)
```

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt [hunk 8]
```diff
@@ -20,18 +21,22 @@ import org.wordpress.android.util.AppLog.T.READER
 */
 object ReaderDiscoverServiceStarter {
     const val ARG_DISCOVER_TASK = "discover_task"
+    const val ARG_SORTING_TYPE = "sortingType"
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt]:
```suggestion
    const val ARG_SORTING_TYPE = "sorting_type"
```

**Review by ashiagr — CHANGES_REQUESTED** [review]
Thanks for working on it @kaung-myat-min! Changes looks nice overall, its great that you could understand the code and put all the pieces together 👍.

I've left few comments/questions/warnings for you to consider. Most are minor, marked some with ⚠️ which can either lead to a crash or alter the behaviour.

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt (hunk 1 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt]: Fixed in 15bba09

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt (hunk 3 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt]: Fixed in 15bba09

File: WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverViewModel.kt (hunk 4 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverViewModel.kt]: Fixed in 860c8f3

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt (hunk 7 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt]: Fixed in 609d151

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt (hunk 8 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverServiceStarter.kt]: Fixed in 609d151

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt (hunk 5 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt:309]: Fixed in 18bbeb3

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt (hunk 2 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt]: @kaung-myat-min, thanks for adding missing comma in 15bba09 but I now see that semi-colon at the end is giving a syntax error.

`Result: near ";": syntax error`

 ```
" $CARDS_SORTING_TYPE_COLUMN TEXT DEFAULT ${DiscoverSortingType.NONE.sortedBy};" +
")"
```
Can we make this small correction by placing it after the closing bracket on the next line?

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt [hunk 9]
```diff
@@ -298,12 +295,18 @@ class ReaderDiscoverLogic(
         ReaderDiscoverCardsTable.addCardsPage(simplifiedCardsJson.toString(), sortingType)
     }
 
+    /**
+     * Clear the saved data
+     *
+     * - Delete the Blogs that are not subscribed / interested by user
+     * - Delete card sequences for given sorting type
+     * - Posts are not deleted because they are used in other sorting type and they will be deleted on purging.
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt:303]: Thank you for this!

File: WordPress/src/main/java/org/wordpress/android/util/config/FilterDiscoverFeatureConfig.kt [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Configuration of the filter for discover or topic feature.
+ */
+@FeatureInDevelopment
+class FilterDiscoverFeatureConfig
+@Inject constructor(appConfig: AppConfig) : FeatureConfig(
+        appConfig,
+        BuildConfig.ENABLE_FILTER_FOR_DISCOVER,
+        FILTER_DISCOVER
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/util/config/FilterDiscoverFeatureConfig.kt]: While feature is in development, we don't need this third param `FILTER_DISCOVER`. 

```
* @param remoteField is the key of the feature flag in the remote config file, only set this field when there is a
* remote flag available
```

Let's just have below in the config file: 

```
@FeatureInDevelopment
class FilterDiscoverFeatureConfig
@Inject constructor(appConfig: AppConfig) : FeatureConfig(
        appConfig,
        BuildConfig.ENABLE_FILTER_FOR_DISCOVER
) 
```

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt (hunk 9 above)
Comment (KaungMyat-Min) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt:303]: Glad, it is useful.

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt (hunk 2 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/datasets/ReaderDiscoverCardsTable.kt]: Fixed in ce8343c

**Pull request merged**